### ISSUE TITLE
Error when using node

### ISSUE THREAD

**Issue by CapsE** [bmad4ever/comfyui_quilting#3]
Hi, when using your nodes (both latent and image) I get this error

Error occurred when executing ImageQuilting_Bmad:

A task has failed to un-serialize. Please ensure that the arguments of the function are all picklable.

File "C:\ComfyUI\ComfyUI\execution.py", line 151, in recursive_execute
output_data, output_ui = get_output_data(obj, input_data_all)
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
File "C:\ComfyUI\ComfyUI\execution.py", line 81, in get_output_data
return_values = map_node_over_list(obj, input_data_all, obj.FUNCTION, allow_interrupt=True)
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
File "C:\ComfyUI\ComfyUI\execution.py", line 74, in map_node_over_list
results.append(getattr(obj, func)(**slice_dict(input_data_all, i)))
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
File "C:\ComfyUI\ComfyUI\custom_nodes\comfyui_quilting\nodes.py", line 158, in compute
texture = quilt_single_with_parallelization(
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
File "C:\ComfyUI\ComfyUI\custom_nodes\comfyui_quilting\nodes.py", line 21, in quilt_single_with_parallelization
return generate_texture_parallel(src, block_size, overlap, outH, outW, tolerance, parallelization_lvl, rng,
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
File "C:\ComfyUI\ComfyUI\custom_nodes\comfyui_quilting\parallel_quilting.py", line 152, in generate_texture_parallel
stripes = Parallel(n_jobs=4, backend="loky", timeout=None)(
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
File "C:\ComfyUI\python_embeded\Lib\site-packages\joblib\parallel.py", line 1952, in __call__
return output if self.return_generator else list(output)
^^^^^^^^^^^^
File "C:\ComfyUI\python_embeded\Lib\site-packages\joblib\parallel.py", line 1595, in _get_outputs
yield from self._retrieve()
File "C:\ComfyUI\python_embeded\Lib\site-packages\joblib\parallel.py", line 1699, in _retrieve
self._raise_error_fast()
File "C:\ComfyUI\python_embeded\Lib\site-packages\joblib\parallel.py", line 1734, in _raise_error_fast
error_job.get_result(self.timeout)
File "C:\ComfyUI\python_embeded\Lib\site-packages\joblib\parallel.py", line 736, in get_result
return self._return_or_raise()
^^^^^^^^^^^^^^^^^^^^^^^
File "C:\ComfyUI\python_embeded\Lib\site-packages\joblib\parallel.py", line 754, in _return_or_raise
raise self._result

**Comment by bmad4ever** [comment]
Hi, @CapsE ,

Thank you for reaching out, and sharing the error stack trace.

I will first share how **to avoid (not solve)** the problem as of the time of writing:
1. Set the option "parallelization level" to zero. 
2. It may be the case that you get a similar, but slightly different, error when using an image batch, because, albeit not yet documented, images in the batch are processed in parallel. If this is the case use lists instead of image batches.
3. If you find that parallelization level zero is too slow for the job, you can try to install the same custom nodes from the "feature-make-seamless" branch (it will require openCV). Using the new "version" option set to 1 and parallelization level to zero, you should be able to run without the error and generate the texture considerably faster.

Now, regarding the problem:
The error is raised at the start of the texture generation, when generating the initial stripes in parallel.
My initial guess is that the problematic arg(s) are the arrays with image data (but I could be wrong). 
If this guess is correct, maybe the way to fix it is to use shared memory.
**I can't, however, reproduce this error, likely to differences in the backend, so unless someone volunteers to solve the problem I will postpone the fix.** 
Once I'm done with the "feature-make-seamless" branch, not being able to use the joblib parallelization will become less of a problem, as the new solution is considerably faster, so I will remain focused on that for now.

Kind Regards

**Comment by bmad4ever** [comment]
@CapsE ,

Since you can reproduce the error, could you investigate the problematic arguments and post them here for future reference?

The arguments are set here: 
https://github.com/bmad4ever/comfyui_quilting/blob/300d6dda80bf12f41158541342f49d5e7f32ec58/parallel_quilting.py#L145

You can also modify the functions "fill_column" and "fill_row" to return None immediately, to check if the joblib "completes" that step instead of the funcs raising an error due to missing data.
https://github.com/bmad4ever/comfyui_quilting/blob/300d6dda80bf12f41158541342f49d5e7f32ec58/parallel_quilting.py#L49

Your help would be much appreciated.

**Comment by CapsE** [comment]
I can confirm that setting parallelization_lvl to 0 fixes the issue but it leads to runtimes of upwards of 45min for a single image in my case.

I'm no Python developer so sadly I can't make any sense of the args. It's a whole bunch of arrays of floats. This is what just the "image" variable prints to console:

[[[     0.2549     0.32549     0.19608]
  [    0.25882     0.34902     0.16471]
  [    0.26667     0.36471     0.16863]
  ...
  [    0.32549     0.46275     0.20784]
  [    0.34118      0.4549     0.21569]
  [    0.37255     0.41569     0.24706]]

 [[    0.25882     0.37255     0.17647]
  [    0.25882     0.37647     0.15686]
  [    0.26667         0.4     0.16863]
  ...
  [    0.32549     0.47059     0.21176]
  [    0.33333     0.46275     0.20784]
  [    0.34902     0.44706     0.21569]]

 [[    0.23529     0.38039     0.16863]
  [    0.27059     0.41569     0.18824]
  [    0.28235      0.4549         0.2]
  ...
  [    0.32549     0.47059     0.20784]
  [    0.32549     0.46275         0.2]
  [    0.33333     0.46275     0.20784]]

 ...

 [[    0.21961     0.36471     0.14118]
  [    0.21961     0.35686     0.14118]
  [    0.21961     0.36471      0.1451]
  ...
  [    0.21569     0.33725     0.16863]
  [    0.21961     0.33725     0.17255]
  [    0.22353     0.31765     0.18039]]

 [[    0.22745     0.37647     0.15686]
  [    0.21961     0.37255     0.14118]
  [    0.22745     0.37647     0.14902]
  ...
  [    0.20392     0.32549     0.15686]
  [    0.21176     0.32549     0.16078]
  [    0.21569      0.2902     0.16471]]

 [[    0.23922     0.33725     0.18824]
  [    0.22353     0.34902     0.16471]
  [    0.21961     0.36863     0.16471]
  ...
  [    0.20784     0.31373     0.17647]
  [    0.20392     0.29412     0.15686]
  [    0.24314      0.2902     0.18431]]]

All the other variables follow the same pattern with different numbers. Not sure if this is expected. 

Returning none immediatly within fill_column and fill_row has no effect on the error.

**Comment by bmad4ever** [comment]
@CapsE , thank you for reporting back,

#### Regarding the generation time:
45 minutes is a lot. I forgot to mention in the previous post, **but you can also improve the speed by increasing the block_size**. Increasing the block size, and using the new branch (supposing you did not try it), you might be able to reduce the generation time to less than 10 minutes. 


#### Concerning the error investigation:
Judging by the time of generation, I thought that numpy could be handling large arrays differently and tried reproducing the error w/ a large image (3840x2160) but w/ no success.
I'm starting to think the problem might be related to the random generator...

If not too much to ask, could you:
1. share the resolution of the image you are using as input;
2. try replacing the lines 145 to 153 with the following variants and report back which variant reached the last print statement? 
(mind that, if you are using the "seamless" branch  there will be an extra argument named "version" and the location of the args might be in some other line)
P.S. **Almost forgot, sending None will raise an error when executing the functions fill_row, and fill_column; like before, you should set them to return None immediately.**

Variant A
```
    args = [
        (image, start_block, overlap, cols_per_quad, tolerance, None),
        (hi_image, hi_start_block, overlap, cols_per_quad, tolerance, None),
        (image, start_block, overlap, rows_per_quad, tolerance, None),
        (vi_image, vi_start_block, overlap, rows_per_quad, tolerance, None)
    ]
    funcs = [fill_row, fill_row, fill_column, fill_column]
    stripes = Parallel(n_jobs=4, backend="loky", timeout=None)(
        delayed(funcs[i])(*args[i]) for i in range(4))
    print("generator was the problem")
```

Variant B
```
    args = [
        (None, None, overlap, cols_per_quad, tolerance, rng),
        (None, None, overlap, cols_per_quad, tolerance, rng),
        (None, None, overlap, rows_per_quad, tolerance, rng),
        (None, None, overlap, rows_per_quad, tolerance, rng)
    ]
    funcs = [fill_row, fill_row, fill_column, fill_column]
    stripes = Parallel(n_jobs=4, backend="loky", timeout=None)(
        delayed(funcs[i])(*args[i]) for i in range(4))
    print("image data was the problem.")
```

Variant C
```
    args = [
        (None, None, overlap, cols_per_quad, tolerance, None),
        (None, None, overlap, cols_per_quad, tolerance, None),
        (None, None, overlap, rows_per_quad, tolerance, None),
        (None, None, overlap, rows_per_quad, tolerance, None)
    ]
    funcs = [fill_row, fill_row, fill_column, fill_column]
    stripes = Parallel(n_jobs=4, backend="loky", timeout=None)(
        delayed(funcs[i])(*args[i]) for i in range(4))
    print("both were the problem.")
```

**Comment by CapsE** [comment]
I'm using a 512x512 image.

All three code variants lead to the same error and the print command is never reached. I installed this node using the node-manager from within ComfyUI but cloning this repo directly and even switching to the other branch didn't help in resolving this issue. I also made sure to load the workflow by dropping the text image from your readme onto ComfyUI to load it's workflow.

This is the full error message. I think it was always the same in every case but I didn't look super close.

!!! Exception during processing!!! A task has failed to un-serialize. Please ensure that the arguments of the function are all picklable.
joblib.externals.loky.process_executor._RemoteTraceback:
"""
Traceback (most recent call last):
  File "C:\ComfyUI\python_embeded\Lib\site-packages\joblib\externals\loky\process_executor.py", line 426, in _process_worker
    call_item = call_queue.get(block=True, timeout=timeout)
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "multiprocessing\queues.py", line 122, in get
ModuleNotFoundError: No module named 'comfyui_quilting'
"""

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "C:\ComfyUI\ComfyUI\execution.py", line 151, in recursive_execute
    output_data, output_ui = get_output_data(obj, input_data_all)
                             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\ComfyUI\ComfyUI\execution.py", line 81, in get_output_data
    return_values = map_node_over_list(obj, input_data_all, obj.FUNCTION, allow_interrupt=True)
                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\ComfyUI\ComfyUI\execution.py", line 74, in map_node_over_list
    results.append(getattr(obj, func)(**slice_dict(input_data_all, i)))
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\ComfyUI\ComfyUI\custom_nodes\comfyui_quilting\nodes.py", line 158, in compute
    texture = quilt_single_with_parallelization(
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\ComfyUI\ComfyUI\custom_nodes\comfyui_quilting\nodes.py", line 21, in quilt_single_with_parallelization
    return generate_texture_parallel(src, block_size, overlap, outH, outW, tolerance, parallelization_lvl, rng,
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\ComfyUI\ComfyUI\custom_nodes\comfyui_quilting\parallel_quilting.py", line 152, in generate_texture_parallel
    stripes = Parallel(n_jobs=4, backend="loky", timeout=None)(
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\ComfyUI\python_embeded\Lib\site-packages\joblib\parallel.py", line 1952, in __call__
    return output if self.return_generator else list(output)
                                                ^^^^^^^^^^^^
  File "C:\ComfyUI\python_embeded\Lib\site-packages\joblib\parallel.py", line 1595, in _get_outputs
    yield from self._retrieve()
  File "C:\ComfyUI\python_embeded\Lib\site-packages\joblib\parallel.py", line 1699, in _retrieve
    self._raise_error_fast()
  File "C:\ComfyUI\python_embeded\Lib\site-packages\joblib\parallel.py", line 1734, in _raise_error_fast
    error_job.get_result(self.timeout)
  File "C:\ComfyUI\python_embeded\Lib\site-packages\joblib\parallel.py", line 736, in get_result
    return self._return_or_raise()
           ^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\ComfyUI\python_embeded\Lib\site-packages\joblib\parallel.py", line 754, in _return_or_raise
    raise self._result
joblib.externals.loky.process_executor.BrokenProcessPool: A task has failed to un-serialize. Please ensure that the arguments of the function are all picklable.

**Comment by bmad4ever** [comment]
@CapsE , Thank you for going the extra mile and testing the code.

Now I can rule out the size of the data as a potential problem.

Joblib will always raise an exception, because it is also used later in the code. However, the stack trace you shared is the same as the previous one, the problem still occurs when generating the stripes.
```
File "C:\ComfyUI\ComfyUI\custom_nodes\comfyui_quilting\parallel_quilting.py", line 152, in generate_texture_parallel
stripes = Parallel(n_jobs=4, backend="loky", timeout=None)(
```

**The error stack trace you just shared was obtained when running variant C?**
If this is indeed the case I'm stumped for now.

**Comment by bmad4ever** [comment]
I finally reproduced this error locally using a fresh portable comfyui.
The error is quite misleading as it appears to have nothing to do with the pickability of the function.

**The problem seems to be that the python module is not found, which can be solved by placing a symlink of `comfyui_quilting` inside the `python_embeded\Lib\site-packages`.**
But, I don't think I can implement this hack via code as it would require admin permissions for the symlink.
I could do a copy instead, but it would be a somewhat dirty solution.

Separating the quilting code from the comfy nodes into another repo could work and would allow usage in other applications; however, the nodes also use parallelization when using batches so this solution requires a bit of refactoring.
I will likely explore this avenue once I get the time.

I'm not aware of other potential solutions at the time of writing.


(update -> even w/ separated projects the reference to the comfy custom node folder is still required... or maybe I'm missing something here??? this might take some time to investigate properly... )

(update 2 -> solved locally, for parallel lvl, needs refactoring for batch behavior and then a new repo)

**Comment by bmad4ever** [comment]
It should have been fixed now by #9 ; if not, please re-open the issue.
